controller/sync.go

```diff
@@ -539,3 +547,41 @@ func syncWindowPreventsSync(app *v1alpha1.Application, proj *v1alpha1.AppProject
 	}
 	return !window.CanSync(isManual)
 }
+
+const impersonateUserNameFormat = "system:serviceaccount:%s"
+const defaultServiceAccountName = "default"
+
+// setImpersonationConfig sets the impersonation config if the feature is enabled via environment variable explicitly.
+func setImpersonationConfig(cfg *rest.Config, app *v1alpha1.Application, proj *v1alpha1.AppProject) {
+	serviceAccountName := deriveServiceAccountName(proj, app)
+	cfg.Impersonate = rest.ImpersonationConfig{
+		UserName: fmt.Sprintf(impersonateUserNameFormat, serviceAccountName),
+	}
+}
+
+// deriveServiceAccountName determines the service account to be used for impersonation for the application sync operation.
+func deriveServiceAccountName(project *v1alpha1.AppProject, application *v1alpha1.Application) string {
+	// spec.Destination.Namespace is optional. If not specified, use the Application's
+	// namespace
+	serviceAccountNamespace := application.Namespace
+	if application.Spec.Destination.Namespace != "" {
+		serviceAccountNamespace = application.Spec.Destination.Namespace
+	}
+	// Loop through the destinationServiceAccounts and see if there is any destination that is an exact match
+	// if so, return the service account specified for that destination.
```

Comment:
The destination can be a glob match too right? So maybe reword it to not say exact match?

Reply:
Corrected it with `candidate` instead of exact match.